executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,449 @@
+	joiner            joiner
+	joinChkResourceCh chan *chunk.Chunk
+	resultCh          chan *indexHashJoinResult
+	taskCh            <-chan *indexHashJoinTask
+	wg                *sync.WaitGroup
+	finished          atomic.Value
+}
+
+type indexHashJoinResult struct {
+	chk *chunk.Chunk
+	err error
+	src chan<- *chunk.Chunk
+}
+
+type outerRowStatusFlag byte
+
+const (
+	_ outerRowStatusFlag = iota // outerRowUnmatched
+	outerRowMatched
+	outerRowHasNull
+)
+
+type indexHashJoinTask struct {
+	*lookUpJoinTask
+	outerRowStatus []outerRowStatusFlag
+	err            error
+}
+
+// Open implements the IndexNestedLoopHashJoin Executor interface.
+func (e *IndexNestedLoopHashJoin) Open(ctx context.Context) error {
+	// Be careful, very dirty hack in this line!!!
+	// IndexLookUpJoin need to rebuild executor (the dataReaderBuilder) during
+	// executing. However `executor.Next()` is lazy evaluation when the RecordSet
+	// result is drained.
+	// Lazy evaluation means the saved session context may change during executor's
+	// building and its running.
+	// A specific sequence for example:
+	//
+	// e := buildExecutor()   // txn at build time
+	// recordSet := runStmt(e)
+	// session.CommitTxn()    // txn closed
+	// recordSet.Next()
+	// e.dataReaderBuilder.Build() // txn is used again, which is already closed
+	//
+	// The trick here is `getStartTS` will cache start ts in the dataReaderBuilder,
+	// so even txn is destroyed later, the dataReaderBuilder could still use the
+	// cached start ts to construct DAG.
+	_, err := e.innerCtx.readerBuilder.getStartTS()
+	if err != nil {
+		return err
+	}
+
+	err = e.children[0].Open(ctx)
+	if err != nil {
+		return err
+	}
+	e.memTracker = memory.NewTracker(e.id, e.ctx.GetSessionVars().MemQuotaIndexLookupJoin)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+	e.innerPtrBytes = make([][]byte, 0, 8)
+	e.startWorkers(ctx)
+	return nil
+}
+
+func (e *IndexNestedLoopHashJoin) startWorkers(ctx context.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupJoinConcurrency
+	workerCtx, cancelFunc := context.WithCancel(ctx)
+	e.cancelFunc = cancelFunc
+	innerCh := make(chan *indexHashJoinTask, concurrency)
+	e.workerWg.Add(1)
+	ow := e.newOuterWorker(innerCh)
+	go util.WithRecovery(func() { ow.run(workerCtx) }, e.finishJoinWorkers)
+
+	e.resultCh = make(chan *indexHashJoinResult, concurrency+1)
+	e.joinChkResourceCh = make([]chan *chunk.Chunk, concurrency)
+	for i := int(0); i < concurrency; i++ {
+		e.joinChkResourceCh[i] = make(chan *chunk.Chunk, 1)
+		e.joinChkResourceCh[i] <- newFirstChunk(e)
+	}
+	e.workerWg.Add(concurrency)
+	for i := int(0); i < concurrency; i++ {
+		workerID := i
+		go util.WithRecovery(func() { e.newInnerWorker(innerCh, workerID).run(workerCtx) }, e.finishJoinWorkers)
+	}
+	go e.wait4JoinWorkers()
+}
+
+func (e *IndexNestedLoopHashJoin) finishJoinWorkers(r interface{}) {
+	if r != nil {
+		e.resultCh <- &indexHashJoinResult{err: errors.Errorf("%v", r)}
```

Comment:
select `ctx.Done` here?

Reply:
why?

Reply:
If we do not block the worker here when context cancels, we do not need to wait for `resultCh` in `Close` method.

Reply:
But `select ctx.Done` here seems do not help?
The workflow here is:
```
outer or inner worker panic (main thread does not know)
  --> goroutine is recovered 
      --> send the error message to resultCh 
          --> main thread gets the error and return the error
              --> Close is called
                  --> ctx.Done is called
```

Reply:
Ok, waiting for `resultCh` is ok for me.